### ISSUE TITLE
How can I compare two index definitions?

### ISSUE THREAD

**Issue by aihornmac** [rethinkdb/rethinkdb#5990]
I am trying to implement a soft-create mechanism. When it comes to index, I wonder how to compare two index definitions to decide whether or not to override. I noticed `indexStatus` returns query and function, but they don't match `[any query].toString()`. I wonder if there is a more direct way to tell if two queries do the same thing.

**Comment by danielmewes** [comment]
I don't think this is currently possible. You could theoretically evaluate the query string that you get from `indexStatus` in JavaScript, and then compare the resulting query ASTs. You'll need to be careful to ignore things like variable numbers though.

Can you explain a bit more what your usage scenario looks like? Maybe there's another way to solve it.

**Comment by marshall007** [comment]
@aihornmac I've needed to do the same thing in the past. As @danielmewes mentioned you can pretty easily normalize both queries before comparing them with something like this:

```js
let replace = {}
let index = 0

query.replace(/_var(\d+)/g, (match, id) => {
  if (replace[id] == null) replace[id] = index++
  return `_var${replace[id]}`
})
```

Another trick is to create the indices you want to compare against on an empty temp table.  You can then compare the binary data in the `function` field you get back from `.indexStatus()` between the two. I've found this approach to be more reliable than parsing the strings.

**Comment by aihornmac** [comment]
@danielmewes We have this init script run every time there are changes on some certain tables in development process (same in deployment). The script mocks fake data (in dev, not in prod), add or recreate (if they can be compared) indexes.

**Comment by aihornmac** [comment]
@marshall007 Sometimes the query sent and the query from `indexStatus` look differently (if var names not considered) even if they do the same thing, so regular expression replacement might be inaccurate. However the latter workaround seem pretty reliable, yet I am not sure if there could be any side effects or not? @danielmewes Could you evaluate if there would be problems?

**Comment by danielmewes** [comment]
Comparing the binary index function should work, but whenever you get different string representations in `indexStatus`, you're also going to have different binary representations. The string representation is deterministically generated from the binary one.

Another thing to keep in mind is that the binary representation I think is different depending on which version of RethinkDB the index was created with (because it contains a version tag unless I'm mistaken). So if you want to be able to compare indexes across RethinkDB versions, using the string will likely work better.

@aihornmac Do you have an example where two semantically identical index functions look different in the printed `indexStatus` string?

**Comment by danielmewes** [comment]
@aihornmac Would it be an option to generate the index names in a certain consistent way, so that you can identify identical indexes based on their names alone?
